processor/k8sattributesprocessor/internal/kube/client.go

```diff
@@ -374,12 +374,26 @@ func (c *WatchClient) extractPodAttributes(pod *api_v1.Pod) map[string]string {
 	return tags
 }
 
-func (c *WatchClient) extractPodContainersAttributes(pod *api_v1.Pod) map[string]*Container {
-	containers := map[string]*Container{}
-
-	if c.Rules.ContainerImageName || c.Rules.ContainerImageTag {
+func (c *WatchClient) extractPodContainersAttributes(pod *api_v1.Pod) PodContainers {
+	containers := PodContainers{
+		ByID:   map[string]*Container{},
+		ByName: map[string]*Container{},
+	}
+	if !needContainerAttributes(c.Rules) {
+		return containers
+	}
+	ifRule := func(rule bool, v string) string {
+		if rule {
+			return v
+		}
+		return ""
+	}
+	switch {
+	case c.Rules.ContainerImageName,
```

Comment:
Side-note (@dmitryax): I think complexity of this code (and overall) could be greatly reduced if we would _always_ set all fields of these structs, and only check the rules when setting the actual attributes. It seems like concerns would be better separated that way and the complexity here reduced. WDYT? I suspect it wasn't memory that was a concern here. If it was, that can still be addressed (better and) differently.